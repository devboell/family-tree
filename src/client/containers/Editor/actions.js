/* eslint-disable import/prefer-default-export */
import {
  SET_CREATE_MODE,
} from './actionTypes'


export const setCreateMode = flag => ({
  type: SET_CREATE_MODE,
  flag,
})
