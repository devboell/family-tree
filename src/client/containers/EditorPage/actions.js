import {
  SET_SELECTED_PERSON_ID,
  SET_CREATE_MODE,
  SET_IS_REMOVING,
} from './constants'

export const setSelectedPersonId = id => ({
  type: SET_SELECTED_PERSON_ID,
  id,
})

export const setCreateMode = flag => ({
  type: SET_CREATE_MODE,
  flag,
})

export const setIsRemoving = flag => ({
  type: SET_IS_REMOVING,
  flag,
})

export const clearSelection = () => ({
  type: SET_SELECTED_PERSON_ID,
  id: 'no_selection',
})
