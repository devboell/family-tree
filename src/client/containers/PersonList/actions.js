import {
  SET_SELECTED_PERSON_ID,
  SET_SELECTION_MODE,
} from './actionTypes'

export const setSelectedPersonId = id => ({
  type: SET_SELECTED_PERSON_ID,
  id,
})

export const setSelectionMode = mode => ({
  type: SET_SELECTION_MODE,
  mode,
})
