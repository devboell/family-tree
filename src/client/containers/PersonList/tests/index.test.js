import React from 'react'
import { defaults } from 'lodash/fp'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { PersonList, mapDispatchToProps, mapStateToProps } from '../index'
import selectionModes from '../constants'
import { setSelectedPersonId, setSelectionMode } from '../actions'

const MockChild = () => <div />
const persons = [{ id: '3' }, { id: '4' }]


describe('PersonList component', () => {
  let defaultProps

  beforeEach(() => {
    defaultProps = {
      loading: false,
      persons: [],
      refetch: jest.fn(),
      selectedPersonId: 'no_selection',
      selectionMode: selectionModes.SELECT_FIRST,
      onSelectPerson: jest.fn(),
      onSelectFirst: jest.fn(),
      onUnselect: jest.fn(),
    }
  })
  describe('LOADING is true', () => {
    let props
    beforeEach(() => {
      props = defaults(defaultProps)({ loading: true })
    })

    it('shows loading ... message', () => {
      const wrapper = shallow(
        <PersonList {...props}>
          <MockChild />
        </PersonList>)

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('list is EMPTY', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(
        <PersonList {...defaultProps}>
          <MockChild />
        </PersonList>)
    })

    it('renders empty list', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('displays empty list message', () => {
      const emptyMessageText = wrapper.find('EmptyListMessage').dive().text()
      expect(emptyMessageText).toBe('no persons found')
    })

    it('passes empty person to child component', () => {
      const person = wrapper.find('MockChild').props().person
      expect(person).toEqual({})
    })
  })

  // tests the logic in componentWillReceiveProps
  describe('list receives persons prop, selectionMode is SELECT_FIRST', () => {
    let wrapper

    beforeEach(() => {
      wrapper = shallow(
        <PersonList {...defaultProps}>
          <MockChild />
        </PersonList>)
    })

    it('calls onSelectPerson with id of first list item', () => {
      wrapper.setProps({ persons })
      expect(defaultProps.onSelectPerson).toHaveBeenCalledWith('3')
    })

    it('does not call onSelectPerson when persons is empty', () => {
      wrapper.setProps({ persons: [] })
      expect(defaultProps.onSelectPerson).not.toHaveBeenCalled()
    })
  })

  describe('list is NOT EMPTY, selectionMode is SELECTED', () => {
    let props

    beforeEach(() => {
      props = defaults(defaultProps)(
        {
          persons,
          selectionMode: selectionModes.SELECTED,
          selectedPersonId: '3',
        },
      )
    })

    it('renders', () => {
      const wrapper = shallow(
        <PersonList {...props}>
          <MockChild />
        </PersonList>)

      expect(wrapper).toMatchSnapshot()
    })

    it('calls onSelectPerson when Button is clicked', () => {
      const wrapper = shallow(
        <PersonList {...props}>
          <MockChild />
        </PersonList>)
      const button = wrapper.find('Button').at(1)
      button.simulate('click')
      expect(props.onSelectPerson).toHaveBeenCalledWith('4')
    })
  })
})

describe('PersonList mapDispatchToProps, mapStateToProps', () => {
  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore({})

  const {
    onSelectPerson,
    onSelectFirst,
    onUnselect,
  } = mapDispatchToProps(store.dispatch)

  beforeEach(() => {
    store.clearActions()
  })

  it('maps onSelectPerson to selectPerson thunk', () => {
    const id = '1'
    const expectedActions = [
      setSelectedPersonId(id),
      setSelectionMode(selectionModes.SELECTED),
    ]
    onSelectPerson(id)
    const receivedActions = store.getActions()
    expect(receivedActions).toEqual(expectedActions)
  })

  it('maps onSelectFirst to selectFirst thunk', () => {
    const expectedActions = [
      setSelectedPersonId('no_selection'),
      setSelectionMode(selectionModes.SELECT_FIRST),
    ]
    onSelectFirst()
    const receivedActions = store.getActions()
    expect(receivedActions).toEqual(expectedActions)
  })

  it('maps onUnselect to unselect thunk', () => {
    const expectedActions = [
      setSelectedPersonId('no_selection'),
      setSelectionMode(selectionModes.UNSELECTED),
    ]
    onUnselect()
    const receivedActions = store.getActions()
    expect(receivedActions).toEqual(expectedActions)
  })

  it('maps state to props', () => {
    const state = {
      personList: {
        selectedPersonId: '2',
        selectionMode: selectionModes.SELECTED,
      },
    }

    const stateProps = mapStateToProps(state)
    expect(stateProps.selectedPersonId).toBe('2')
    expect(stateProps.selectionMode).toBe(selectionModes.SELECTED)
  })
})
