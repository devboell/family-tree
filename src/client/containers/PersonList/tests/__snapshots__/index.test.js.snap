// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonList component LOADING is true shows loading ... message 1`] = `
<styled.div>
  loading ...
</styled.div>
`;

exports[`PersonList component list is EMPTY renders empty list 1`] = `
<styled.div>
  <styled.div>
    <EmptyListMessage>
      no persons found
    </EmptyListMessage>
  </styled.div>
  <styled.div>
    <MockChild
      onSelectFirst={[Function]}
      onSelectPerson={[Function]}
      onUnselect={[Function]}
      person={Object {}}
      personId="no_selection"
      persons={Array []}
      refetch={[Function]}
      selectionMode="SELECT_FIRST"
    />
  </styled.div>
</styled.div>
`;

exports[`PersonList component list is NOT EMPTY, selectionMode is SELECTED renders 1`] = `
<styled.div>
  <styled.div>
    <styled.ul>
      <styled.li>
        <Button
          isSelected={true}
          onClick={[Function]}
        />
      </styled.li>
      <styled.li>
        <Button
          isSelected={false}
          onClick={[Function]}
        />
      </styled.li>
    </styled.ul>
  </styled.div>
  <styled.div>
    <MockChild
      onSelectFirst={[Function]}
      onSelectPerson={[Function]}
      onUnselect={[Function]}
      person={
        Object {
          "id": "3",
        }
      }
      personId="3"
      persons={
        Array [
          Object {
            "id": "3",
          },
          Object {
            "id": "4",
          },
        ]
      }
      refetch={[Function]}
      selectionMode="SELECTED"
    />
  </styled.div>
</styled.div>
`;
